source/utils/date.cpp

```diff
@@ -96,6 +85,11 @@ bool date::operator!=(const date &comparand) const
 
 int date::to_number(calendar base_date) const
 {
+    if (_is_null)
+    {
+        throw xlnt::exception("cannot convert invalid/empty date to a number");
```

Comment:
Most of the time a derived class is thrown. `xlnt::invalid_attribute` may be appropriate.

Reply:
I agree. For better exception handling using `try` / `catch`, using a derived exception class would be better. However, at the moment, using `xlnt::exception` allows me to set an exception message, while using `xlnt::invalid_attribute` does not. Since I don't like using `try` / `catch` all over my code (see my long explanation above), I would personally prefer having a proper error message instead of a derived exception class - but I know that other programmers might disagree with me.

We would have the following options:

1. Adapt `xlnt::invalid_attribute` to accept a custom error message, meaning that both advantages could be achieved at once.
2. Switch to `xlnt::invalid_attribute` for a better `try` / `catch` experience, at the cost of a worse error message.
3. Leave the exceptions how they are at the moment, with good error messages, but not so great `try` / `catch` experience.

Which one of the options should we choose? :slightly_smiling_face: I would personally prefer option 1 (while keeping the current default message, if no custom one is specified), but I'm open for discussion :wink: 

Note: I just pushed a commit with even better error messages when throwing exceptions - however, I can gladly switch to any other option.

Reply:
Option 1 is fine for me.

Reply:
My last commit contains your suggested improvements :wink: